src/nccl_ofi_net.c

```diff
@@ -1414,40 +1416,73 @@ static ncclResult_t ofi_connect(int dev, void *handle, void **sendComm)
 	}
 
 	/* Send "connect" message to remote EP */
-	do {
-		/*
-		 * TODO: replace it with API of FI_INJECT type when most of
-		 * providers can support it, so that need for completion check
-		 * below can be lifted.
-		 */
-		rc = fi_tsend(sComm->local_ep, (void *)&local_ep_addr,
-			      MAX_EP_ADDR, NULL, sComm->remote_ep,
-			      sComm->tag | ~max_tag, &req->ctx);
-		if (rc == 0)
-			break;
+	/* do not wait for completion in case of sending to self */
+	if (memcmp(local_ep_addr, remote_ep_addr, namelen) == 0) {
+		static char buffer_to_self[MAX_EP_ADDR];
+		NCCL_OFI_TRACE(NCCL_NET, "Connect to self");
+		if (request_to_self != NULL) {
```

Comment:
Here, the implementation assumes that there could only be a single `request_to_self` at any point. NCCL is a multi-threaded application which establishes connection with multiple peers (including GPUs within node). We would possibly like to make this a part of sendComm structure.

Reply:
Working on this.

Reply:
Sorry, if we make it part of sendComm structure, we will not be able to access it from ofi_accept() as it does not take sendComm as a parameter.

Reply:
Well, you could send the pointer to request from `tsend` so that accept side can free it once it gets that.

Reply:
Reworked